src/jit/lsra.cpp

```diff
@@ -5671,12 +5686,42 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
         {
             RegRecord* physRegToUnassign = availablePhysRegInterval;
 #ifdef _TARGET_ARM_
-            // We should unassign a double register if availablePhysRegInterval is part of the double register
-            if (availablePhysRegInterval->assignedInterval->registerType == TYP_DOUBLE &&
-                !genIsValidDoubleReg(availablePhysRegInterval->regNum))
-                physRegToUnassign = findAnotherHalfRegRec(availablePhysRegInterval);
+            RegRecord* anotherPhysRegToUnassign  = nullptr;
+            Interval*  anotherIntervalToUnassign = nullptr;
+
+            if (currentInterval->registerType == TYP_DOUBLE)
+            {
+                if (intervalToUnassign->registerType == TYP_FLOAT)
+                {
+                    // We should unassign not only physRegToUnassign which is a float register
+                    // but also another half float register if the currentInterval has a double register.
+                    // So, additional variables would be set.
+                    anotherPhysRegToUnassign = findAnotherHalfRegRec(physRegToUnassign);
+                    assert(anotherPhysRegToUnassign != nullptr);
+                    anotherIntervalToUnassign = anotherPhysRegToUnassign->assignedInterval;
+                    assert(intervalToUnassign != anotherIntervalToUnassign);
+                }
+            }
+            else // TYP_FLOAT
+            {
+
+                // We should unassign a double register if availablePhysRegInterval is part of the double register
+                if (intervalToUnassign->registerType == TYP_DOUBLE &&
+                    !genIsValidDoubleReg(availablePhysRegInterval->regNum))
+                {
+                    physRegToUnassign = findAnotherHalfRegRec(availablePhysRegInterval);
+                }
+            }
 #endif
             unassignPhysReg(physRegToUnassign, intervalToUnassign->recentRefPosition);
```

Comment:
This is the point at which, for `_TARGET_ARM_`, you need to discover whether there is an additional `Interval` that needs to be unassigned. I would suggest that perhaps for `_TARGET_ARM_` it would be more efficient to defer the identification of `intervalToUnassign`, and instead check whether `(bestScore & VALUE_AVAILABLE) == 0` in which case we know there is at least one `Interval` to unassign. Then `unassignPhysReg()` would have a different signature that would simply take the `RegRecord` and the `RegisterType`. Then, if we are unassigning a `TYP_DOUBLE`, we would unassigned both halves, and otherwise if the `intervalToUnassign` is `TYP_DOUBLE` then too we would unassign both halves. I think that would be cleaner, even though it makes the implementation of `unassignPhysReg()` more complex (and more target-dependent).